### PR TITLE
index is included in memory usage by default

### PR DESCRIPTION
**Pull request by max-sixty**
...and sys.getsizeof returns correct value. Closes https://github.com/pydata/pandas/issues/11597

Is this the best implementation for a common method across classes?

### CODE REVIEW

File: pandas/core/base.py [hunk 1]
```diff
@@ -544,6 +544,8 @@ def memory_usage(self, deep=False):
             v += lib.memory_usage_of_objects(self.values)
         return v
 
+    __sizeof__ = com._sizeof
```
Comment (jreback) [inline pandas/core/base.py]: you can put this in ``PandasObject`` I think

File: pandas/core/categorical.py [hunk 2]
```diff
@@ -950,6 +950,8 @@ def memory_usage(self, deep=False):
         """
         return self._codes.nbytes + self._categories.memory_usage(deep=deep)
 
+    __sizeof__ = com._sizeof
```
Comment (jreback) [inline pandas/core/categorical.py]: then you don't need this

File: pandas/core/common.py [hunk 3]
```diff
@@ -2854,6 +2854,16 @@ def pprint_thing_encoded(object, encoding='utf-8', errors='replace', **kwds):
     value = pprint_thing(object)  # get unicode representation of object
     return value.encode(encoding, errors, **kwds)
 
+def _sizeof(self):
+    """
+    Generates the total memory usage for a object that returns either a value or
+    NDFrame of values
+    """
+    mem = self.memory_usage()
```
Comment (jreback) [inline pandas/core/common.py]: use ``not lib.isscalar``, then just return (don' need an ``if/else``)

``int(mem.sum())`` should be enough

File: pandas/core/frame.py [hunk 4]
```diff
@@ -1744,6 +1744,8 @@ def memory_usage(self, index=False, deep=False):
                              index=['Index']).append(result)
         return result
 
+    __sizeof__ = com._sizeof
```
Comment (jreback) [inline pandas/core/frame.py]: actually I would just define this here directly

(instead of in ``com``)

File: doc/source/whatsnew/v0.18.0.txt [hunk 5]
```diff
@@ -40,6 +40,7 @@ Other enhancements
 - ``read_excel`` now supports s3 urls of the format ``s3://bucketname/filename`` (:issue:`11447`)
 - A simple version of ``Panel.round()`` is now implemented (:issue:`11763`)
 - For Python 3.x, ``round(DataFrame)``, ``round(Series)``, ``round(Panel)`` will work (:issue:`11763`)
+- sys.getsizeof(obj) returns the memory usage of the object including (non-object) values (:issue:`11597`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.0.txt]: this is not True at all. that's the purpose of ``deep``. This just allows ``sys.getsizeof(pandas_object)`` to work

File: pandas/core/frame.py (hunk 4 above)
Comment (jreback) [inline pandas/core/frame.py]: and FYI you don't need to define this in sub-classes (you already defined this in ``base``). What I mean here is remove ALL definition of ``__sizeof__`` except the one on ``PandasObject``, and handle the case where the object doesn't have ``.memory_usage`` defined.

File: pandas/tests/test_frame.py [hunk 6]
```diff
@@ -7693,6 +7693,10 @@ def test_info_memory_usage(self):
         DataFrame(1,index=pd.MultiIndex.from_product([['a'],range(1000)]),columns=['A']).index.nbytes
         DataFrame(1,index=pd.MultiIndex.from_product([['a'],range(1000)]),columns=['A']).index.values.nbytes
 
+        # sys.getsizeof works the same as memory_usage with defaults, albeit with some
+        # GC overhead
+        self.assertAlmostEqual(sys.getsizeof(df), df.memory_usage().sum(), delta=100)
```
Comment (jreback) [inline pandas/tests/test_frame.py]: test this on ``series/Index/Categorical`` as well. (for ``Index`` put a test in test_index on the ``Base`` which will run for all indexes).

File: doc/source/whatsnew/v0.18.0.txt [hunk 7]
```diff
@@ -137,6 +138,7 @@ other anchored offsets like ``MonthBegin`` and ``YearBegin``.
 Other API Changes
 ^^^^^^^^^^^^^^^^^
 
+- memory_usage now includes values in the index (:issue:`11597`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.0.txt]: ``.memory_usage`` and mention that ``.info`` will include the memory used by the index as well

File: doc/source/whatsnew/v0.18.0.txt (hunk 5 above)
Comment (MaximilianR) [inline doc/source/whatsnew/v0.18.0.txt]: That's why it states `(non-object)`?
Regardless I'll rephrase it to make it crystal

File: pandas/tests/test_frame.py (hunk 6 above)
Comment (MaximilianR) [inline pandas/tests/test_frame.py]: I added onto the existing tests in `IndexOps` - lmk if that's not OK

**Comment by MaximilianR** [comment]
@jreback Python 2.6 error with `delta` argument on `assertAlmostEqual` - should I skip the test on 2.6 or find an alternate way of doing this?

xref https://github.com/pydata/pandas/issues/7718

File: pandas/tests/test_base.py [hunk 8]
```diff
@@ -883,26 +883,24 @@ def get_fill_value(obj):
...
-                    self.assertTrue(res2 > res)
-                else:
-                    self.assertEqual(res, res2)
-
                 self.assertEqual(o.memory_usage(index=False) + o.index.memory_usage(),
                                  o.memory_usage(index=True))
 
+            # sys.getsizeof will call the .memory_usage defaults, and add on some GC overhead
+            self.assertAlmostEqual(res, sys.getsizeof(o), delta=100)
+
```
Comment (jreback) [inline pandas/tests/test_base.py]: we don't use this

use tm.assert_almost_equal

Comment (jreback) [inline pandas/tests/test_base.py]: I thought we disabled the test routines that one shouldn't use - in tm.TestCase
but maybe that's an open issue (or if not can you create one)

we are dropping 2.6 soon - but still like for there to be 1 way to do testing

Comment (MaximilianR) [inline pandas/tests/test_base.py]: OK - although the functionality is materially worse than the nosetests one (actually kinda peculiar with the bool argument for 3 vs 5): 

>check_less_precise : bool, default False
    Specify comparison precision.
    5 digits (False) or 3 digits (True) after decimal points are compared.

**Comment by MaximilianR** [comment]
@jreback `assertLess` isn't supported either on 2.6? 
But we're green

File: doc/source/whatsnew/v0.18.0.txt [hunk 9]
```diff
@@ -108,6 +108,9 @@ Other enhancements
 - A simple version of ``Panel.round()`` is now implemented (:issue:`11763`)
 - For Python 3.x, ``round(DataFrame)``, ``round(Series)``, ``round(Panel)`` will work (:issue:`11763`)
 - ``Dataframe`` has gained a ``_repr_latex_`` method in order to allow for automatic conversion to latex in a ipython/jupyter notebook using nbconvert. Options ``display.latex.escape`` and ``display.latex.longtable`` have been added to the configuration and are used automatically by the ``to_latex`` method.(:issue:`11778`)
+- sys.getsizeof(obj) returns the memory usage of a pandas object including the (non-object)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.0.txt]: actually I think this  *should* call with ``deep=True``.

File: doc/source/whatsnew/v0.18.0.txt [hunk 10]
```diff
@@ -205,6 +208,7 @@ other anchored offsets like ``MonthBegin`` and ``YearBegin``.
 Other API Changes
 ^^^^^^^^^^^^^^^^^
 
+- .memory_usage now includes values in the index, as does memory_usage in .info (:issue:`11597`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.0.txt]: use double-backticks around ``.info``

File: pandas/tests/test_base.py [hunk 11]
```diff
@@ -883,26 +883,24 @@ def get_fill_value(obj):
 
 
     def test_memory_usage(self):
+        import sys
```
Comment (jreback) [inline pandas/tests/test_base.py]: import sys at the top of the file

File: pandas/tests/test_categorical.py [hunk 12]
```diff
@@ -1211,6 +1211,7 @@ def test_nbytes(self):
         self.assertEqual(cat.nbytes, exp)
 
     def test_memory_usage(self):
+        import sys
```
Comment (jreback) [inline pandas/tests/test_categorical.py]: move imports to the top

File: doc/source/whatsnew/v0.18.0.txt (hunk 9 above)
Comment (MaximilianR) [inline doc/source/whatsnew/v0.18.0.txt]: I don't have enough context to know whether `deep` should be True or False by default, but I can see the logic behind the default for `.memory_usage` should being the same as that for `sys.getsizeof`, given the similar use cases.

Or why are the use cases different? Because a user calling from the system wants accuracy vs. speed but a user calling from pandas wants speed vs. accuracy?

Comment (jreback) [inline doc/source/whatsnew/v0.18.0.txt]: using ``__sizeof__`` *should* have ``deep=True`` which gives the 'best' report of memory used. This is not the default for ``.memory_usage()`` because its expensive to compute (potentially). see https://github.com/pydata/pandas/issues/11595 where it can be somewhat slow (though the cython impl helps).

Comment (jreback) [inline doc/source/whatsnew/v0.18.0.txt]: e.g. if someone is calling ``sys.getsizeof(df)`` then I think its appropriate to give the most accurate (if maybe somewhat non-performant) answer.

Comment (MaximilianR) [inline doc/source/whatsnew/v0.18.0.txt]: OK, if it's a material performance drag and a different enough set of use cases, sobeit

**Comment by MaximilianR** [comment]
@jreback updated & green

File: pandas/core/base.py [hunk 13]
```diff
@@ -112,6 +112,25 @@ def _reset_cache(self, key=None):
...
+        """
+        Generates the total memory usage for a object that returns either a value or
+        Series of values
+        """
+        if hasattr(self, 'memory_usage'):
+            # we want to default to deep=True, but not all object offer that
+            # we could check the kwarg with inspect module, but different methods for Py versions
+            try:
+                mem = self.memory_usage(deep=True)
+            except TypeError:
```
Comment (jreback) [inline pandas/core/base.py]: nothing should raise a ``TypeError``, what does?

**Comment by MaximilianR** [comment]
@jreback updated & green

**Comment by jreback** [comment]
@MaximilianR lgtm.

can you run: ``git diff master | flake8 --diff`` and fix those issues. we are going to be shortly enforcing this, so might as well not add to it :)

ping when green.

**Comment by MaximilianR** [comment]
Done (phew...)

File: pandas/tests/test_categorical.py [hunk 14]
```diff
@@ -1,22 +1,22 @@
...
 
-from pandas import Categorical, Index, Series, DataFrame, PeriodIndex, Timestamp, CategoricalIndex
-
-from pandas.core.config import option_context
-import pandas.core.common as com
+import pandas as pd
 import pandas.compat as compat
+import pandas.core.common as com
 import pandas.util.testing as tm
+from pandas import Categorical, Index, Series, DataFrame, \
```
Comment (jreback) [inline pandas/tests/test_categorical.py]: ok, though we usually do this with parens

**Comment by jreback** [comment]
lgtm. ping when green.

File: pandas/tests/test_frame.py [hunk 15]
```diff
@@ -7697,23 +7697,35 @@ def test_info_memory_usage(self):
...
+                [['a'], range(1000)]),
+            columns=['A']
+        )
+        df.index.nbytes
+        df.memory_usage(index=True)
+        df.index.values.nbytes
+
+        # sys.getsizeof will call the .memory_usage with
+        # deep=True, and add on some GC overhead
+        diff = df.memory_usage().sum() - sys.getsizeof(df)
```
Comment (jreback) [inline pandas/tests/test_frame.py]: how did this pass before?

this will never be true. as ``deep=True`` is >> non-deep with object.

Comment (MaximilianR) [inline pandas/tests/test_frame.py]: Hmmm good point, I will edit

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/60cacabdd37dcfb00920aed453787e77358219d1

thanks!
